### ISSUE TITLE
Nondeterministic typing error on module load

### ISSUE THREAD

**Issue by fritzo** [pyro-ppl/funsor#243]
I have seen a couple nondeterministic errors on loading the funsor library with error messages
```
KeyError: (typing.Union[funsor.ops.LogAddExpOp, funsor.ops.NullOp], <class 'funsor.ops.AddOp'>, <class 'frozenset'>, typing.Tuple[typing.Union[funsor.torch.Tensor, funsor.terms.Number], funsor.gaussian.Gaussian])
```
This is tough to reproduce.

<details>

```sh
$ pytest -vx test/examples/test_bart.py --pdb
============================================== test session starts ==============================================
platform darwin -- Python 3.7.2, pytest-5.0.1, py-1.8.0, pluggy-0.12.0 -- /Users/fritzobermeyer/miniconda2/envs/pyro3/bin/python
cachedir: .pytest_cache
benchmark: 3.2.2 (defaults: timer=time.perf_counter disable_gc=False min_rounds=5 min_time=0.000005 max_time=1.0 calibration_precision=10 warmup=False warmup_iterations=100000)
rootdir: /Users/fritzobermeyer/github/pyro-ppl/funsor, inifile: setup.cfg
plugins: xdist-1.26.1, nbval-0.9.1, forked-1.0.2, benchmark-3.2.2
collecting ...
>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>> traceback >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>
test/examples/test_bart.py:5: in <module>
    import funsor
funsor/__init__.py:8: in <module>
    from . import (
funsor/adjoint.py:7: in <module>
    from funsor.cnf import Contraction, nullop
funsor/cnf.py:170: in <module>
    Tuple[Union[Tensor, Number], Gaussian]]
funsor/terms.py:232: in __getitem__
    return cls._type_cache[arg_types]
../../../miniconda2/envs/pyro3/lib/python3.7/weakref.py:137: in __getitem__
    o = self.data[key]()
E   KeyError: (typing.Union[funsor.ops.LogAddExpOp, funsor.ops.NullOp], <class 'funsor.ops.AddOp'>, <class 'frozenset'>, typing.Tuple[typing.Union[funsor.torch.Tensor, funsor.terms.Number], funsor.gaussian.Gaussian])
>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>> entering PDB >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>

>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>> PDB post_mortem (IO-capturing turned off) >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>
> /Users/fritzobermeyer/miniconda2/envs/pyro3/lib/python3.7/weakref.py(137)__getitem__()
-> o = self.data[key]()
(Pdb)
collected 0 items

=============================================== warnings summary ================================================
funsor/terms.py:6
  /Users/fritzobermeyer/github/pyro-ppl/funsor/funsor/terms.py:6: DeprecationWarning: Using or importing the ABCs from 'collections' instead of from 'collections.abc' is deprecated, and in 3.8 it will stop working
    from collections import Hashable, OrderedDict

-- Docs: https://docs.pytest.org/en/latest/warnings.html
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!! _pytest.outcomes.Exit: Quitting debugger !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
========================================== 1 warnings in 97.33 seconds ==========================================
```

</details>

**Comment by eb8680** [comment]
Hmm, looks like it's related to `weakref` and garbage collection or to nondeterminism in importing?  We could probably work around this by adding a bit of error handling in `FunsorMeta.__getitem__`

**Comment by fritzo** [comment]
@eb8680 did you fix this in #329 ?

**Comment by eb8680** [comment]
I don't know, since I can't reliably reproduce this, but it's possible.
